### DISCUSSION TITLE
How to integrate StrongSwan with MPTCP?

### DISCUSSION THREAD

**Discussion by daniel0720** [strongswan/strongswan#2460] [Q&A]
Hello everyone,

I am looking to integrate StrongSwan with MPTCP. Could anyone provide some advice and guidance on this?

My experimental environment consists of two terminals, each with two network interfaces that can communicate with each other.

Currently, I have conducted two experiments:

An MPTCP experiment where I successfully aggregated the bandwidth of two paths on both terminals.

A StrongSwan multipath experiment where I established two IPsec tunnels between the two terminals. However, these two tunnels do not work simultaneously; instead, one takes over when the other fails.

I want to make these two tunnels work simultaneously, with MPTCP utilizing both tunnels for data transmission. Could you please advise on how to configure StrongSwan for this purpose?

Thank you.

**Comment by tobiasbrunner** [comment]
> A StrongSwan multipath experiment where I established two IPsec tunnels between the two terminals. However, these two tunnels do not work simultaneously; instead, one takes over when the other fails.

Could you please clarify what you configured exactly and what you observed? Also, have you tried using two [XFRM interfaces](https://docs.strongswan.org/docs/5.9/features/routeBasedVpn.html)?

**Reply by daniel0720** [reply]
The topology is as follows.

![image](https://github.com/user-attachments/assets/feee779f-a288-491f-a42e-ec50619f0309)

The moon swanctf.conf is as follows.

![image](https://github.com/user-attachments/assets/6b8a78e2-2e04-4786-b414-deb449cabdb7)

The sun config file is similar.

Then I started strongswan, only 6to6 connection works. The route table 220 has only one route entry. 

I haven't tested using two XFRM interfaces yet. I just tested one XFRM interface. If I use two XFRM interfaces, can both tunnels work simultaneously？Can I specify distinct route table for each XFRM interface? If so, I think I can add two endpoints using xfrm interfaces for MPTCP. 

Thank you very much for your reply.

**Reply by tobiasbrunner** [reply]
> Then I started strongswan, only 6to6 connection works.

Obviously, as you have duplicate traffic selectors configured. That will result in a single outbound policy in the kernel that can direct traffic only to a single IPsec SA (usually the latest that was negotiated). You need marks or interface IDs to install separate policies in order to make use of both SAs. No idea if it's possible to mark parts of the MPTCP connection differently. If not, XFRM interfaces are the way to go.

> If I use two XFRM interfaces, can both tunnels work simultaneously？

Yes.

> Can I specify distinct route table for each XFRM interface?

You can do whatever you want with the routing as strongSwan will not install routes by default if you configure an interface ID.

**Reply by daniel0720** [reply]
Thank you.
according to your device, I use two XFRM interfaces. The configuration of moon is as follows.
1. add two xfrm interfaces: xfrm0 and xfrm1
2. add two routes via xfrm0 and xfrm1 to bob(10.2.0.10)
3. add iptables rules in and out xfrm0 and xfrm1
![image](https://github.com/user-attachments/assets/3c8e40fe-df50-481e-9ed8-cbb0289bba65)

The swanctl.conf of moon is as follows. there are two conns: 5to5 and 6to6. each conn has its if_id. same traffic selector.
![image](https://github.com/user-attachments/assets/b26187ab-c5f6-43a9-a58a-4075a54da1f0)
![image](https://github.com/user-attachments/assets/384819c0-701b-4bdd-a6ac-d1f1780f5597)

The swanctl.conf of sun is as follows. there are two conns: 5to5 and 6to6. each conn has its if_id. same traffic selector. For the convience, I set fixed interface ID. Note that the updown script is different.
![image](https://github.com/user-attachments/assets/cdca2323-fff2-41d6-8457-aa5cd0ec16f7)
![image](https://github.com/user-attachments/assets/8ca7cb4a-65f1-4ebe-8da1-fc43c196f04b)

I wrote two updown scripts for each conns: /etc/updown and /etc/updown_1.
/etc/updown
![image](https://github.com/user-attachments/assets/b3da3ba3-2f4d-497c-b054-bc32570bbb88)

/etc/updown_1
![image](https://github.com/user-attachments/assets/4cb7c1a1-1047-4e92-a369-8d99dce0ab12)

according to the log, both 5to5 and 6to6 conns all established. and alice and ping bob through 5to5 conns. 

sun has routes, xfrm interfaces and iptables rules.
![image](https://github.com/user-attachments/assets/afb87145-8cdd-44c9-bcf8-35fe9155869a)
![image](https://github.com/user-attachments/assets/4c48e7f3-4232-4c92-a1c6-61de1aab88a4)
![image](https://github.com/user-attachments/assets/bb6e491f-7985-4a0d-a649-a5054bb097d3)

There should only be 4 iptables rules, but through the logs, it was discovered that moon sent requests twice, resulting in the generation of 4 additional rules.

**but the question is:**  if i terminate 5to5 conn, alice can't ping bob. what I expect is 6to6 conn continues to work.

Is there any config errors I've made? I am new to strongswan. how could I make two conns work simultaneously? or when the one fails, the other continues to work?

Thank you again for your help.

**Reply by tobiasbrunner** [reply]
If you use `start_action = start` on both sides, duplicates are definitely possible (even if you configure a uniqueness policy).

> I wrote two updown scripts for each conns: /etc/updown and /etc/updown_1.

Note that the interface IDs are already passed to the script as environment variables (`PLUTO_IF_ID_IN|OUT`), as are e.g. the traffic selectors. And you can also pass arguments to the script via swanctl.conf (e.g. the metric).

> **but the question is:** if i terminate 5to5 conn, alice can't ping bob. what I expect is 6to6 conn continues to work.

What happens on moon? If the interface and route for 5to5 are still there, why would it send traffic via the other interface and the 6to6 SA? Either you need a script there too or some dynamic routing daemon that verifies the connectivity via the interfaces.

**Reply by daniel0720** [reply]
> Note that the interface IDs are already passed to the script as environment variables (PLUTO_IF_ID_IN|OUT), as are e.g. the traffic selectors. And you can also pass arguments to the script via swanctl.conf (e.g. the metric).

I am sorry but I am new to strongswan and I can't understand how the PLUTO_IF_ID_IN|OUT environment variables work in the updown script. I wrote two updown scripts because I want to control each connection separately. For example, if I want to terminate 5to5 SA, the updown script of sun will not affect 6to6 SA. I don't know if i can do this in one script.

> What happens on moon? 

the moon executes `sudo swanctl --terminate --ike 5to5` command.  yes you are correct. the routes and interface of moon are static, not dynamically installed.

>  Either you need a script there too 

Can I use updown script in the swanctl.conf of moon?

**Reply by tobiasbrunner** [reply]
> I can't understand how the PLUTO_IF_ID_IN|OUT environment variables work in the updown script.

They are just environment variables passed to the script. You can have a look at the [default script](https://github.com/strongswan/strongswan/blob/master/src/_updown/_updown.in) for a list (and some examples on their usage, although they obviously won't fit your use case). For instance, the interface name could just be defined like this:

    IF_NAME_ENS6="xfrm-$PLUTO_IF_ID_OUT"

> For example, if I want to terminate 5to5 SA, the updown script of sun will not affect 6to6 SA.

Due to the variables and the parameters passed to the script, you can easily write a single script to handle both connections.

> Can I use updown script in the swanctl.conf of moon?

Sure, why not?

**Reply by daniel0720** [reply]
I have read the default _updown script, what I don't understand is how the PLUTO_IF_ID_OUT changes. for example, IF_NAME_ENS5 and IF_NAME_ENS6 should be different, increase 1 is better. but I don't know how it works. even more, what's the relationship between PLUTO_IF_ID_OUT environment variables and %unique of swanctl.conf.

>  you can easily write a single script to handle both connections.

thank you, I'll try to test it.

**Reply by tobiasbrunner** [reply]
> what I don't understand is how the PLUTO_IF_ID_OUT changes. for example, IF_NAME_ENS5 and IF_NAME_ENS6 should be different, increase 1 is better

The environment variables are filled with values from the respective CHILD_SA, so they depend on whatever you configure in swanctl.conf (or what's negotiated for some values like traffic selectors).

> even more, what's the relationship between PLUTO_IF_ID_OUT environment variables and %unique of swanctl.conf

With `%unqiue`, each instance of a CHILD_SA created from a config will get a separate identifier. If you configure a static value and two CHILD_SAs are created (you already mentioned that duplicates were created), both would have the same static value assigned. Either way, the value you get in `PLUTO_IF_ID_OUT` is the value determined for the CHILD_SA for which the script is called (either from the config or allocated via `%unqiue`).

**Reply by daniel0720** [reply]
as you suggest, I wrote a updown script in moon, as follows.
![image](https://github.com/user-attachments/assets/39b2acd0-83b5-44f2-8c94-a910e0565901)

And changed the swanctl.conf, start_action = trap
![image](https://github.com/user-attachments/assets/6b5b1b27-dbef-4b79-ae3a-351ef26a1e97)

Then I `swanctl --load-all`, `swanctl --initiate --ike 5to5`, the log is as follows.
![image](https://github.com/user-attachments/assets/9ede705c-23ec-4ecd-a88b-229334969ca3)

However, it seems that the updown script has not been executed. No xfrm interface added, no routes added , no iptables rules added.

**Reply by tobiasbrunner** [reply]
> as you suggest, I wrote a updown script in moon, as follows.

Looks like unnecessarily duplicated code. Instead of hard-coding `ens5` and `ens6` just use `$1` if you pass the name as an argument. You can also pass the metric as a second argument, or determine it based on the interface name. Also note that the underlying interface is optional for XFRM interfaces (it has currently no effect). So you might not need that name (the interface ID is different anyway).




> Then I `swanctl --load-all`, `swanctl --initiate --ike 5to5`, the log is as follows.

You only initiate the IKE_SA that way, the script runs only once the CHILD_SA is established (use `--child <name>` to initiate a child).

**Reply by daniel0720** [reply]
> You only initiate the IKE_SA that way, the script runs only once the CHILD_SA is established (use --child <name> to initiate a child).

Thank you very much. It works after `swanctl --initiate --child n-n`. Then I exec `swanctl --initiate --child net-net`, now two child sa established. One continues to work after the other fails.

But the syslog show that `mptcpd: add_addr: too many addresses or duplicate one: -34`. Why?
![image](https://github.com/user-attachments/assets/c25018ae-18aa-4f4c-9aaa-327d3e183dd6)

How can I use these two child sa to transfer tcp data by mptcp? Could you please give me some advice? Really appreciate.

**Reply by tobiasbrunner** [reply]
Maybe `mptcpd` doesn't like that the interface doesn't have any IP addresses on it (or rather only an IPv6 link-local address). But I really don't know as I've never used mptcp.

**Reply by daniel0720** [reply]
OK，Thank you very much. You helped me a lot.
